File: TPs/TP02/eatery.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

typedef struct {
    int ano;
    int mes;
    int dia;
} Data;

Data parse_data(char* s){
    Data d;

    sscanf(s, "%d-%d-%d", &d.ano, &d.mes, &d.dia);

    return d;
}

void formatar_data(Data* d, char* buffer){
    sprintf(buffer, "%02d/%02d/%04d", d->dia, d->mes, d->ano);
}

typedef struct {
    int hora;
    int minuto;
} Hora;

Hora parse_hora(char* s){
    Hora h;

    sscanf(s, "%d:%d", &h.hora, &h.minuto);

    return h;
}

void formatar_hora(Hora* h, char* buffer){
    sprintf(buffer, "%02d:%02d", h->hora, h->minuto);
}

typedef struct {
    int id;
    char* nome;
    char* cidade;
    int capacidade;
    double avaliacao;
    int n_tipos_cozinha;
    char** tipos_cozinha;
    int faixa_preco;
    Hora horario_abertura;
    Hora horario_fechamento;
    Data data_abertura;
    bool aberto;

} Restaurante;

Restaurante* parse_restaurante(char* s){
    
    Restaurante* r = (Restaurante*) malloc(sizeof(Restaurante));
    char parte[300];
    int p_parte = 0;

    int coluna = 0;

    for (int i = 0;  s[i] != '\0'; i++){

        if (s[i] == ','){
            parte[p_parte] = '\0';

            if (coluna == 0) { //id
                sscanf(parte, "%d", &r->id);
            } else if (coluna == 1){ //nome
                int tam = 0;
                while (parte[tam] != '\0') tam++;

                r->nome = (char*) malloc((tam+1) * sizeof(char));
                strcpy(r->nome, parte);
            } else if (coluna == 2){ //cidade
                int tam = 0;
                while (parte[tam] != '\0') tam++;
                
                r->cidade = (char*) malloc((tam+1) * sizeof(char));
                strcpy(r->cidade, parte);
            } else if (coluna == 3){ //capacidade
                sscanf(parte, "%d", &r->capacidade);
            } else if (coluna == 4){ //avaliacao
                sscanf(parte, "%lf", &r->avaliacao);
            } else if (coluna == 5){ //tipos_cozinha
                //conta qtd de palavras pra criar vetor
                int qtd = 1;
                for (int i = 0; parte[i] != '\0'; i++){
                    if (parte[i] == ';') qtd++;
                }

                r->n_tipos_cozinha = qtd;

                //alloca vetor de char(string) de vetor de char (string) com o tamanho de cozinha
                char** cozinha = (char**) malloc(qtd * sizeof(char*));

                //varre e guarda letras com vetor temporário


                char tmp[500];
                int pos_tmp = 0;
                int tipo_atual = 0;

                for (int j = 0; parte[j] != '\0'; j++){
                    char letra = parte[j];

                    if (letra == ';'){
                        tmp[pos_tmp] = '\0';

                        cozinha[tipo_atual] = (char*) malloc((pos_tmp + 1) * sizeof(char));
                        
                        for (int k = 0; k <= pos_tmp; k++){
                            cozinha[tipo_atual][k] = tmp[k];
                        }

                        tipo_atual++;
                        pos_tmp = 0;
                    } else {
                        tmp[pos_tmp] = letra;
                        pos_tmp++;
                    }
                }

                tmp[pos_tmp] = '\0';
                cozinha[tipo_atual] = (char*) malloc((pos_tmp+1) * sizeof(char));
            
                for (int j = 0; j <= pos_tmp; j++){
                    cozinha[tipo_atual][j] = tmp[j];
                }

                r->tipos_cozinha = cozinha;

            } else if (coluna == 6){ //faixa_preco
                int preco = 0;
                for (int i = 0; parte[i] != '\0'; i++) preco++;

                r->faixa_preco = preco;
            } else if (coluna == 7){ //horario
            char tmp_ab[6];
            char tmp_fe[6];
            int j = 0;
            
            while (parte[j] != '-'){
                tmp_ab[j] = parte[j];
                j++;
            }
            tmp_ab[j] = '\0';
            
            j++;
            
            int k = 0;
            while (parte[j] != '\0'){
                tmp_fe[k] = parte[j];
                j++;
                k++;
            }
            tmp_fe[k] = '\0';
            
            r->horario_abertura = parse_hora(tmp_ab);
            r->horario_fechamento = parse_hora(tmp_fe);
            } else if (coluna == 8){ //data_abertura
                r->data_abertura = parse_data(parte);    
            }
            coluna++;
            p_parte = 0;
        } else {
            if (s[i] != '\n' && s[i] != '\r') {
                parte[p_parte] = s[i];
                p_parte++;
            }
        }
    }
    //aberto 
    parte[p_parte] = '\0';
    r->aberto = (strcmp(parte, "true") == 0) ? true : false;

    return r;
}

void formatar_restaurante(Restaurante* r, char* buffer){

    //formatando tipos de cozinha
    char tmp[50];
    int pos_tmp = 0;
    for(int i = 0; i < r->n_tipos_cozinha; i++){
        for(int j = 0; r->tipos_cozinha[i][j] != '\0'; j++){
            tmp[pos_tmp] = r->tipos_cozinha[i][j];
            pos_tmp++;
        }
        if((i + 1) < r->n_tipos_cozinha){
            tmp[pos_tmp] = ',';
            pos_tmp++;
        } else {
            tmp[pos_tmp] = '\0';
        }
    }
    int cont = 0;
    for(int i = 0; tmp[i] != '\0'; i++) cont++;

    char cozinhas_formatada[cont + 1];
    for(int i = 0; i < cont; i++){
        cozinhas_formatada[i] = tmp[i];
    }
    cozinhas_formatada[cont] = '\0';
    
    //formatando faixa para '$'
    int tam = r->faixa_preco;
    char faixa_formatada[tam + 1];
    for(int i = 0; i < tam; i++){
        faixa_formatada[i] = '$';
    }
    faixa_formatada[tam] = '\0';

    char abertura_formatado[6];
    char fechamento_formatado[6];
    char data_formatada[11];

    formatar_hora(&r->horario_abertura, abertura_formatado);
    formatar_hora(&r->horario_fechamento, fechamento_formatado);
    formatar_data(&r->data_abertura, data_formatada);


    sprintf(buffer, "[%d ## %s ## %s ## %d ## %.1lf ## [%s] ## %s ## %s-%s ## %s ## %s]",
            r->id,
            r->nome,
            r->cidade,
            r->capacidade,
            r->avaliacao,
            cozinhas_formatada,
            faixa_formatada,
            abertura_formatado,
            fechamento_formatado,
            data_formatada,
            r->aberto ? "true" : "false"
    );
}

typedef struct{
    int tamanho;
    Restaurante** restaurantes;
} Colecao_Restaurantes;

void ler_csv_colecao(Colecao_Restaurantes* colecao, char* path) {
    FILE* arquivo = fopen(path, "r");
    
    if (arquivo != NULL) {
        char linha[1000];

        fgets(linha, 1000, arquivo);

        while (fgets(linha, 1000, arquivo) != NULL) {
            
            Restaurante* r = parse_restaurante(linha);

            colecao->restaurantes[colecao->tamanho] = r;
            colecao->tamanho++;
        }
        
        fclose(arquivo);
    }
}

Colecao_Restaurantes* ler_csv() {
    Colecao_Restaurantes* colecao = (Colecao_Restaurantes*) malloc(sizeof(Colecao_Restaurantes));
    colecao->tamanho = 0;
    
    colecao->restaurantes = (Restaurante**) malloc(1000 * sizeof(Restaurante*));

    ler_csv_colecao(colecao, "/tmp/restaurantes.csv");

    return colecao;
}

int main() {
    Colecao_Restaurantes* colecao = ler_csv();

    char linha[20];
    bool ler_entrada = true;

    while (ler_entrada && scanf(" %s", linha) == 1) {
        
        if (strcmp(linha, "-1") == 0) {
            ler_entrada = false;
        } else {
            int id_procurado = 0;
            for(int i = 0; linha[i] != '\0'; i++) {
                if(linha[i] >= '0' && linha[i] <= '9') {
                    id_procurado = (id_procurado * 10) + (linha[i] - '0');
                }
            }

            bool achou = false;
            int indice = 0;

            while (!achou && indice < colecao->tamanho) {
                if (colecao->restaurantes[indice]->id == id_procurado) {
                    
                    char buffer_saida[2000];
                    formatar_restaurante(colecao->restaurantes[indice], buffer_saida);
                    printf("%s\n", buffer_saida);
                    
                    achou = true;
                }
                indice++;
            }
        }
    }

    return 0;
}

File: TPs/TP02/Eatery.java
import java.util.*;
import java.io.*;

class Data {
    private int ano;
    private int mes;
    private int dia;

    public Data(){
        this.ano = 0;
        this.mes = 0;
        this.dia = 0;
    }

    public int getAno(){
        return this.ano;
    }
    public void setAno(int ano){
        this.ano = ano;
    }
            
    public int getMes(){
        return this.mes;
    }
    public void setMes(int mes){
        this.mes = mes;
    }

    public int getDia(){
        return this.dia;
    }
    public void setDia(int dia){
        this.dia = dia;
    }

    public static Data parseData(String s){
        Data d = new Data();
    
        int num = 0;
        int etapa = 0;
    
        for (int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if (c == '-'){
                if (etapa == 0){     //ano
                    d.setAno(num);
                } else if (etapa == 1){     //mes
                    d.setMes(num);
                }  
                num = 0;
                etapa++;
            } else {
                int digito = c - '0';
                num = (num * 10) + digito;
            }
        }
    
        d.setDia(num);
    
        return d;    
    }

    public String formatar(){
        return String.format("%02d/%02d/%04d", this.dia, this.mes, this.ano);
    }
}
        
class Hora {
    private int hora;
    private int minuto;

    public Hora(){
        this.hora = 0;
        this.minuto = 0;
    }

    public int getHora(){
        return this.hora;
    }
    public void setHora(int hora){
        if (hora >= 0 && hora <= 23) {
            this.hora = hora;
        }
    }

    public int getMinuto(){
        return this.minuto;
    }
    public void setMinuto(int minuto){
        if (minuto >= 0 && minuto <= 59) {
            this.minuto = minuto;
        }
    }

    public static Hora parseHora(String s){
        Hora h = new Hora();

        int num = 0;

        for (int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(c == ':'){
                h.setHora(num);
                num = 0;
            } else {
                int digito = c - '0';
                num = (num * 10) + digito;
            }
        }
        h.setMinuto(num);
        
        return h;
    }

    public String formatar(){
        return String.format("%02d:%02d", this.hora, this.minuto);
    }
}

class Restaurante {
    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tiposCozinha;
    private int faixaPreco;
    private Hora horarioAbertura;
    private Hora horarioFechamento;
    private Data dataAbertura;
    private boolean aberto;

    public Restaurante(){
        this.id = 0;
        this.nome = "";
        this.cidade = "";
        this.capacidade = 0;
        this.avaliacao = 0;
        this.tiposCozinha = new String[0];
        this.faixaPreco = 0;
        this.horarioAbertura = new Hora();
        this.horarioFechamento = new Hora();
        this.dataAbertura = new Data();
        this.aberto = false;        
    }

    public int getId(){
        return this.id;
    }
    public String getNome(){
        return this.nome;
    }
    public String getCidade(){
        return this.cidade;
    }
    public int getCapacidade(){
        return this.capacidade;
    }
    public double getAvaliacao(){
        return this.avaliacao;
    }
    public String[] getTiposCozinha(){
        return this.tiposCozinha;
    }
    public int getFaixaPreco(){
        return this.faixaPreco;
    }
    public Hora getHorarioAbertura(){
        return this.horarioAbertura;
    }
    public Hora getHorarioFechamento(){
        return this.horarioFechamento;
    }
    public Data getDataAbertura(){
        return this.dataAbertura;
    }
    public boolean getAberto(){
        return this.aberto;
    }


    public void setId(int id){
        this.id = id;
    }
    public void setNome(String nome){
        this.nome = nome;
    }
    public void setCidade(String cidade){
        this.cidade = cidade;
    }
    public void setCapacidade(int capacidade){
        this.capacidade = capacidade;
    }
    public void setAvaliacao(double avaliacao){
        this.avaliacao = avaliacao;
    }
    public void setTiposCozinha(String[] tiposCozinha){
        this.tiposCozinha = tiposCozinha;
    }
    public void setFaixaPreco(int faixaPreco){
        if (faixaPreco >= 1 && faixaPreco <= 4){
            this.faixaPreco = faixaPreco;
        }
    }
    public void setHorarioAbertura(Hora horarioAbertura){
        this.horarioAbertura = horarioAbertura;
    }
    public void setHorarioFechamento(Hora horarioFechamento){
        this.horarioFechamento = horarioFechamento;
    }
    public void setDataAbertura(Data dataAbertura){
        this.dataAbertura = dataAbertura;
    }
    public void setAberto(boolean aberto){
        this.aberto = aberto;
    }

    public static Restaurante parseRestaurante(String s){
        Restaurante r = new Restaurante();
        String parte = "";
        int coluna = 0; 
    
        for (int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if (c == ','){
                if (coluna == 0){ //id
                    int num = 0;
                    for (int j = 0; j < parte.length(); j++){
                        char cr = parte.charAt(j);
                        num = (num * 10) + (cr - '0'); 
                    }
                    r.setId(num);
                } else if (coluna == 1){ //nome
                    r.setNome(parte);
                } else if (coluna == 2){ //cidade
                    r.setCidade(parte);
                } else if (coluna == 3){ //capacidade
                    int num = 0;
                    for (int j = 0; j < parte.length(); j++){
                        char cr = parte.charAt(j);
                        num = (num * 10) + (cr - '0'); 
                    }
                    r.setCapacidade(num);                    
                } else if (coluna == 4){ //avaliacao
                    double num = 0.0;
                    double divisor = 10.0;
                    boolean ponto = false;
                    for (int j = 0; j < parte.length(); j++){
                        char cr = parte.charAt(j);
                        if(cr == '.'){
                            ponto = true;
                        } else {
                            if (!ponto){
                                num = (num * 10) + (cr - '0');
                            } else {
                                num = num + ((cr - '0')/ divisor);
                                divisor *= 10;
                            }
                        }
                    }
                    
                    r.setAvaliacao(num);

                } else if (coluna == 5){ //tipos_cozinha

                    int tam = 1;
                    for (int j = 0; j < parte.length(); j++){
                        if(parte.charAt(j) == ';') tam++;
                    }

                    String[] cozinhas = new String[tam]; 

                    int cont = 0;
                    String tipo = "";

                    for (int j = 0; j < parte.length(); j++){
                        char cr = parte.charAt(j);

                        if (cr == ';'){
                            cozinhas[cont] = tipo;
                            cont++;
                            tipo = "";
                        } else {
                            tipo += cr;
                        }
                    }
                    cozinhas[cont] = tipo;

                    r.setTiposCozinha(cozinhas);
                } else if (coluna == 6){ //faixa_preco
                    int faixa = parte.length();
                    r.setFaixaPreco(faixa);
                } else if (coluna == 7){ //horario
                    Hora abertura = new Hora();
                    String hr = "";
                    for (int j = 0; j < parte.length(); j++){
                        char cr = parte.charAt(j);
                        
                        if(cr == '-'){
                            abertura = Hora.parseHora(hr);
                            hr = "";
                        } else {
                            hr += cr;
                        }
                    }
                    Hora fechamento = Hora.parseHora(hr);

                    r.setHorarioAbertura(abertura);
                    r.setHorarioFechamento(fechamento);
                } else if (coluna == 8){ //data_abertura
                    Data d = Data.parseData(parte);
                    r.setDataAbertura(d);
                } //fim if
                coluna++;
                parte = "";
            } else {
                parte += s.charAt(i);
            } // fim if ','
        } // fim for
        
        boolean aberto = false;

        if (parte.compareTo("true") == 0){
            aberto = true;
        }    
        
        r.setAberto(aberto);

        return r;
    }// fim parseRestaurante

    public String formatar(){
        // formatando os tipos de cozinha e faixa de preço antes de formar a string geral
        String cozinhasFormatadas = "[";
        for (int i = 0; i < this.tiposCozinha.length; i++){
            cozinhasFormatadas += tiposCozinha[i];
            if (i < tiposCozinha.length - 1) cozinhasFormatadas += ",";
        }
        cozinhasFormatadas += "]";

        String precoFormatado = "";
        for (int i = 0; i < this.faixaPreco; i++){
            precoFormatado += "$";
        }

        return String.format("[%d ## %s ## %s ## %d ## %s ## %s ## %s ## %s-%s ## %s ## %s]", 
                                        this.id,
                                        this.nome,
                                        this.cidade,
                                        this.capacidade,
                                        String.valueOf(this.avaliacao),
                                        cozinhasFormatadas,
                                        precoFormatado,
                                        this.horarioAbertura.formatar(),
                                        this.horarioFechamento.formatar(),
                                        this.dataAbertura.formatar(),
                                        this.aberto);
    }
}

class ColecaoRestaurantes {
    private int tamanho;
    private Restaurante[] restaurantes;

    public ColecaoRestaurantes() {
        this.tamanho = 0;
        this.restaurantes = new Restaurante[1000];
    }

    public int getTamanho(){
        return this.tamanho;
    }
    public Restaurante[] getRestaurantes(){
        return this.restaurantes;
    }

    public void lerCsv(String path){
        try {
        java.io.File arquivo = new java.io.File(path);
        java.util.Scanner leitor = new java.util.Scanner(arquivo);

        if (leitor.hasNextLine()) {
            leitor.nextLine();
        }

        while (leitor.hasNextLine()) {
            String linha = leitor.nextLine();
            
            Restaurante r = Restaurante.parseRestaurante(linha);
            
            this.restaurantes[this.tamanho] = r;
            this.tamanho++;
        }

        leitor.close();
    } catch (java.io.FileNotFoundException e) {
        System.out.println("Arquivo não encontrado!");
    }
    }

    public static ColecaoRestaurantes lerCsv() {
        ColecaoRestaurantes colecao = new ColecaoRestaurantes();
        colecao.lerCsv("/tmp/restaurantes.csv");
        return colecao;
    }


}



public class Eatery{

    public static void ordenarPorInsercao(Restaurante[] array, int n, int[] log) {
        for (int i = 1; i < n; i++) {
            Restaurante tmp = array[i];
            int j = i - 1;
            boolean inseriu = false;

            while (j >= 0 && !inseriu) {
                log[0]++;
                
                if (array[j].getCidade().compareTo(tmp.getCidade()) > 0) {
                    array[j + 1] = array[j];
                    log[1]++;
                    j--;
                } else {
                    inseriu = true; 
                }
            }
            array[j + 1] = tmp;
            log[1]++;
        }
    }

    public static boolean pesquisaSequencial(Restaurante[] array, int n, String nomeBuscado, int[] log) {
        boolean achou = false;
        int i = 0;

        while (i < n && !achou) {
            log[0]++;
            if (array[i].getNome().compareTo(nomeBuscado) == 0) {
                achou = true;
            }
            i++;
        }

        return achou;
    }
    public static void main(String[] args) {
        ColecaoRestaurantes colecao = ColecaoRestaurantes.lerCsv();
        Restaurante[] base = colecao.getRestaurantes();
        Restaurante[] arrayPesquisa = new Restaurante[1000];
        int nPesquisa = 0;
        
        Scanner sc = new Scanner(System.in);

        boolean lerEntrada = true;

        while(lerEntrada && sc.hasNextLine()){
            String linha = sc.nextLine();

            if(linha.compareTo("-1") == 0){
                lerEntrada = false;
            } else {
                //convertendo ID em int
                int idProcurado = 0;
                for(int i = 0; i < linha.length(); i++){
                    char c = linha.charAt(i);
                    if(c >= '0' && c <= '9'){
                        idProcurado = (idProcurado * 10) + (c - '0');
                    }
                }

                boolean achou = false;
                int indice = 0;
                while (!achou && indice < colecao.getTamanho()) {
                    if (base[indice].getId() == idProcurado) {
                        arrayPesquisa[nPesquisa] = base[indice];
                        nPesquisa++;
                        achou = true;
                    }
                    indice++;
                }
            }
        }

        int[] contadoresLog = new int[1];
        long inicioTempo = System.currentTimeMillis();

        boolean lendoNomes = true;
        while (lendoNomes && sc.hasNextLine()) {
            String nomeBuscado = sc.nextLine();

            nomeBuscado = nomeBuscado.replace("\r", "");

            if (nomeBuscado.compareTo("FIM") == 0) {
                lendoNomes = false;
            } else {
                boolean resultado = pesquisaSequencial(arrayPesquisa, nPesquisa, nomeBuscado, contadoresLog);
                
                if (resultado) {
                    System.out.println("SIM");
                } else {
                    System.out.println("NAO");
                }
            }
        }
        sc.close();

        long fimTempo = System.currentTimeMillis();
        long tempoTotal = fimTempo - inicioTempo;

        try {
            FileWriter arquivoLog = new FileWriter("896238_sequencial.txt");
            PrintWriter gravador = new PrintWriter(arquivoLog);
            gravador.printf("896238\t%d\t%d\n", contadoresLog[0], tempoTotal);
            gravador.close();
        } catch (IOException e) {
            System.out.println("Erro ao gravar log");
        }
    }
}